File: src/main/java/fr/lewon/bot/runner/bot/operation/BotOperation.java
package fr.lewon.bot.runner.bot.operation;

import fr.lewon.bot.runner.bot.Bot;
import fr.lewon.bot.runner.bot.props.BotPropertyDescriptor;
import fr.lewon.bot.runner.bot.props.BotPropertyStore;

import java.util.List;

public abstract class BotOperation {

    private final String label;

    public BotOperation(String label) {
        this.label = label;
    }

    public String getLabel() {
        return this.label;
    }

    public abstract List<BotPropertyDescriptor> getNeededProperties(Bot bot);

    public abstract Object run(Bot bot, BotPropertyStore paramsPropertyStore) throws Exception;

}

File: src/main/kotlin/fr/lewon/bot/runner/bot/operation/BotOperation.kt
package fr.lewon.bot.runner.bot.operation

import fr.lewon.bot.runner.Bot
import fr.lewon.bot.runner.bot.props.BotPropertyDescriptor
import fr.lewon.bot.runner.bot.props.BotPropertyStore

abstract class BotOperation(val label: String) {

    abstract fun getNeededProperties(bot: Bot): List<BotPropertyDescriptor>

    @Throws(Exception::class)
    abstract fun run(bot: Bot, paramsPropertyStore: BotPropertyStore): OperationResult

}
